Sources/SecureStorage/SecureStorageDatabaseProvider.swift

```diff
@@ -0,0 +1,169 @@
+//
+//  SecureStorageDatabaseProvider.swift
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import Common
+import GRDB
+
+public protocol SecureStorageDatabaseProvider {
+
+    init(file: URL, key: Data) throws
+
+    static func recreateDatabase(withKey key: Data) throws -> Self
+
+}
+
+open class GRDBSecureStorageDatabaseProvider: SecureStorageDatabaseProvider {
+
+    public enum DatabaseWriterType {
+        case queue
+        case pool
+    }
+
+    /// Provides the default database directory and file name.
+    ///
+    /// This is used to derive the database path relative to the application support directory.
+    open class var databaseLocation: (directoryName: String, fileName: String) {
+        fatalError("Must be overridden by a subclass")
+    }
+
+    /// Determines the GRDB DatabaseWriter type.
+    ///
+    /// The available options are `queue` and `pool`, representing a `DatabaseQueue` and `DatabasePool` respectively.
+    open class var writerType: DatabaseWriterType {
+        fatalError("Must be overridden by a subclass")
+    }
+
+    /// Configures the database migrations to use for the subclass of the database provider.
+    ///
+    /// This is called by the database provider's `init` function as a part of setting up the database.
+    open class func registerMigrations(with migrator: inout DatabaseMigrator) throws {
+        fatalError("Must be overridden by a subclass")
+    }
```

Comment:
Not super happy with this, it's all stuff that should really be provided by the initializer, but because this thing needs to use a really generic initializer thanks to the `SecureStorageDatabaseProvider` protocol then this is what it has. This lets the factory create new instances of the provider without having to know any of this stuff, but it could be improved - open to suggestions.